src/enkf/gridio_fv3reg.f90

```diff
@@ -537,192 +511,182 @@ subroutine writegriddata(nanal1,nanal2,vars3d,vars2d,n3d,n2d,levels,ndim,vargrid
+       workvar3d=workvar3d+workinc3d
+
        call write_fv3_restart_data3d(varstrname,fv3filename,file_id,workvar3d)
        do k=1,nlevs
-          if (nproc .eq. 0)                                               &
+          if (nproc .eq. 0)                                                &
              write(6,*) 'WRITEregional : delp ',                           &
-                 & k, minval(qworkvar3d(:,:,k)), maxval(workvar3d(:,:,k))
+                 & k, minval(workvar3d(:,:,k)), maxval(workvar3d(:,:,k))
        enddo
     end if
 
     if(q_ind>0) then
-    
        varstrname='sphum'
        call read_fv3_restart_data3d(varstrname,fv3filename,file_id,qbgworkvar3d)
-         do k=1,nlevs
-             nn = nn_tile0
-         do j=1,ny_res
-            do i=1,nx_res
-               nn=nn+1
-               workinc3d(i,j,k)=vargrid(nn,levels(q_ind-1)+k,nb,ne) 
-            enddo
-         enddo
-         enddo
-       qworkvar3d=qbgworkvar3d+workinc3d   
-     
+       do k=1,nlevs
+           nn = nn_tile0
+       do j=1,ny_res
+          do i=1,nx_res
+             nn=nn+1
+             workinc3d(i,j,k)=vargrid(nn,levels(q_ind-1)+k,nb,ne)
+          enddo
+       enddo
+       enddo
+       qworkvar3d=qbgworkvar3d+workinc3d
+
        call write_fv3_restart_data3d(varstrname,fv3filename,file_id,qworkvar3d)
        do k=1,nlevs
-          if (nproc .eq. 0)                                               &
+          if (nproc .eq. 0)                                                 &
              write(6,*) 'WRITEregional : sphum ',                           &
                  & k, minval(qworkvar3d(:,:,k)), maxval(qworkvar3d(:,:,k))
        enddo
+       ! During write_fv3_restart_data3d, qworkvar3d got all its dimensions in
+       ! revered order. So, re-calculate qworkvar3d after being written, so
+       ! that it can be used below for the tsen calculation if needed.
+       qworkvar3d=qbgworkvar3d+workinc3d
     end if
 
-
-
-    if (tv_ind > 0.or.tsen_ind>0 ) then
-         
-      varstrname = 'T'
-      if(tsen_ind>0) then
-        do k=1,nlevs
-           nn = nn_tile0
-         do j=1,ny_res
-            do i=1,nx_res
-               nn=nn+1
-               workinc3d(i,j,k)=vargrid(nn,levels(tsen_ind-1)+k,nb,ne) 
-            enddo
-         enddo
-        enddo
-      call read_fv3_restart_data3d(varstrname,fv3filename,file_id,workvar3d)
-      workvar3d=workvar3d+workinc3d
-      call write_fv3_restart_data3d(varstrname,fv3filename,file_id,workvar3d)
-      else  ! tv_ind >0  
-        do k=1,nlevs
-           nn = nn_tile0
-           do j=1,ny_res
+    if (tv_ind>0 .or. tsen_ind>0 ) then
+       varstrname = 'T'
+       if (tsen_ind>0) then
+          do k=1,nlevs
+             nn = nn_tile0
+          do j=1,ny_res
              do i=1,nx_res
                 nn=nn+1
-                workinc3d(i,j,k)=vargrid(nn,levels(tv_ind-1)+k,nb,ne) 
+                workinc3d(i,j,k)=vargrid(nn,levels(tsen_ind-1)+k,nb,ne)
```

Comment:
@TingLei-NOAA I think I also fixed this potential bug when using tsen as an enkf control vector variable. Please help to confirm this is a valid fix. Thanks!

Reply:
Bin, 
From the new codes, yes, it should be tsen_ind used.  However, the comparison blocks in this interface doesn't work well, since many format change interfere .  As shown in this page, the tsen_ind was already used in line 597 in the prior version.   I could n't find where the tv_ind was  (as a mistake) used in the previous version.  Did I miss something?   Or some other versions were used ?

Reply:
Hi @TingLei-NOAA, yes you are right. The original version used the correct tsen_ind. This showed up because of my whitespace and indentation formatting clean-ups. Sorry for the false alarm. Please just ignore my first comment. I am marking it as resolved. Thanks!